File: src/worker/routes/push.ts
import { createRoute, z } from "@hono/zod-openapi";
import { and, eq } from "drizzle-orm";
import {
  PushConfigSchema,
  PushPrefsSchema,
  PushTestResultSchema,
  SubscribeSchema,
  UnsubscribeSchema,
} from "@shared/schemas";
import type { FamEnv } from "../context";
import { schema } from "../db";
import { createApp, jsonContent } from "../lib";
import { pushToUser } from "../push";

const okSchema = z.object({ ok: z.literal(true) });

// SSRF guard (sec review M2): the worker later POSTs to stored endpoints on
// a schedule, so only real browser push services are accepted.
const ALLOWED_PUSH_HOSTS = [
  /(^|\.)fcm\.googleapis\.com$/, // Chrome/Chromium
  /(^|\.)push\.apple\.com$/, // Safari / iOS
  /(^|\.)push\.services\.mozilla\.com$/, // Firefox
  /(^|\.)mozaws\.net$/, // Firefox (legacy autopush)
  /(^|\.)notify\.windows\.com$/, // Edge (WNS)
];

export function isAllowedPushEndpoint(endpoint: string): boolean {
  try {
    const url = new URL(endpoint);
    return (
      url.protocol === "https:" &&
      ALLOWED_PUSH_HOSTS.some((re) => re.test(url.hostname))
    );
  } catch {
    return false;
  }
}

const config = createRoute({
  method: "get",
  path: "/api/push/config",
  tags: ["push"],
  responses: {
    200: jsonContent(PushConfigSchema, "VAPID public key"),
  },
});

const subscribe = createRoute({
  method: "post",
  path: "/api/push/subscribe",
  tags: ["push"],
  description:
    "Register this browser's push subscription for the signed-in caretaker. Upserts by endpoint.",
  request: {
    body: { content: { "application/json": { schema: SubscribeSchema } } },
  },
  responses: {
    200: jsonContent(okSchema, "Stored"),
    400: jsonContent(
      z.object({ error: z.string(), code: z.string().optional() }),
      "Not a recognized push service endpoint",
    ),
  },
});

const unsubscribe = createRoute({
  method: "post",
  path: "/api/push/unsubscribe",
  tags: ["push"],
  request: {
    body: { content: { "application/json": { schema: UnsubscribeSchema } } },
  },
  responses: {
    200: jsonContent(okSchema, "Removed (if it was yours)"),
  },
});

const getPrefs = createRoute({
  method: "get",
  path: "/api/push/prefs",
  tags: ["push"],
  responses: {
    200: jsonContent(PushPrefsSchema, "Notification preferences"),
  },
});

const putPrefs = createRoute({
  method: "put",
  path: "/api/push/prefs",
  tags: ["push"],
  request: {
    body: { content: { "application/json": { schema: PushPrefsSchema } } },
  },
  responses: {
    200: jsonContent(PushPrefsSchema, "Updated preferences"),
  },
});

const test = createRoute({
  method: "post",
  path: "/api/push/test",
  tags: ["push"],
  description: "Send a test notification to all of the caller's devices.",
  responses: {
    200: jsonContent(PushTestResultSchema, "Delivery count"),
  },
});

export const pushApp = createApp<FamEnv>()
  .openapi(config, (c) => {
    return c.json({ publicKey: c.env.VAPID_PUBLIC_KEY }, 200);
  })
  .openapi(subscribe, async (c) => {
    const body = c.req.valid("json");
    if (!isAllowedPushEndpoint(body.endpoint)) {
      return c.json(
        { error: "Unrecognized push service", code: "BAD_ENDPOINT" },
        400,
      );
    }
    const userId = c.var.sessionData.user.id;
    await c.var.db
      .insert(schema.pushSubscription)
      .values({
        familyId: c.var.familyId,
        userId,
        endpoint: body.endpoint,
        p256dh: body.p256dh,
        auth: body.auth,
      })
      .onConflictDoUpdate({
        target: schema.pushSubscription.endpoint,
        set: {
          userId,
          familyId: c.var.familyId,
          p256dh: body.p256dh,
          auth: body.auth,
        },
      });
    return c.json({ ok: true as const }, 200);
  })
  .openapi(unsubscribe, async (c) => {
    const body = c.req.valid("json");
    await c.var.db
      .delete(schema.pushSubscription)
      .where(
        and(
          eq(schema.pushSubscription.endpoint, body.endpoint),
          eq(schema.pushSubscription.userId, c.var.sessionData.user.id),
        ),
      );
    return c.json({ ok: true as const }, 200);
  })
  .openapi(getPrefs, async (c) => {
    const rows = await c.var.db
      .select()
      .from(schema.pushPref)
      .where(
        and(
          eq(schema.pushPref.userId, c.var.sessionData.user.id),
          eq(schema.pushPref.familyId, c.var.familyId),
        ),
      );
    const hours = rows[0]?.feedReminderHours ?? 0;
    return c.json(
      {
        feedReminderHours: (hours === 3 || hours === 4 || hours === 6
          ? hours
          : 0) as 0 | 3 | 4 | 6,
      },
      200,
    );
  })
  .openapi(putPrefs, async (c) => {
    const body = c.req.valid("json");
    await c.var.db
      .insert(schema.pushPref)
      .values({
        userId: c.var.sessionData.user.id,
        familyId: c.var.familyId,
        feedReminderHours: body.feedReminderHours,
        // A new setting starts a fresh observation window.
        lastRemindedAt: null,
      })
      .onConflictDoUpdate({
        target: [schema.pushPref.userId, schema.pushPref.familyId],
        set: {
          feedReminderHours: body.feedReminderHours,
          lastRemindedAt: null,
        },
      });
    return c.json({ feedReminderHours: body.feedReminderHours }, 200);
  })
  .openapi(test, async (c) => {
    const sent = await pushToUser(c.var.db, c.env, c.var.sessionData.user.id, {
      title: "Pjokk",
      body: "Push works on this device ✅",
      url: "/",
    });
    return c.json({ sent }, 200);
  });

File: apps/server/internal/api/push.go
package api

import (
	"context"
	"net/url"
	"strings"

	"github.com/refsdal/pjokk/server/internal/api/gen"
	"github.com/refsdal/pjokk/server/internal/api/middleware"
	dbgen "github.com/refsdal/pjokk/server/internal/db/gen"
	"github.com/refsdal/pjokk/server/internal/push"
)

// This file ports apps/api/src/routes/push.ts (REF §A1 push.ts): the VAPID
// config endpoint, the subscription lifecycle (subscribe/unsubscribe), the
// per-caretaker feed-reminder preference, and a test-send. Every operation
// here is tierFamilyNoAPIKey (api.go's operationAuthTiers) — a push
// subscription is bound to a signed-in BROWSER, not a programmatic pjk_
// bearer caller, matching apps/api/src/app.ts's
// domainBase.use("/api/push/*", rejectApiKey), which is mounted AFTER
// requireFamily: a key resolves fine through auth/tenancy and is refused
// only at this push-specific gate (see push_test.go's
// TestPushRoutesForbidAPIKeyAuth).

// allowedPushHosts is the SSRF guard apps/api/src/routes/push.ts's sec
// review comment (M2) describes: the cron scheduler (a later task) POSTs to
// stored endpoints unattended, so only real browser push services are ever
// accepted as a subscription target.
var allowedPushHosts = []string{
	"fcm.googleapis.com",        // Chrome/Chromium
	"push.apple.com",            // Safari / iOS
	"push.services.mozilla.com", // Firefox
	"mozaws.net",                // Firefox (legacy autopush)
	"notify.windows.com",        // Edge (WNS)
}

// isAllowedPushEndpoint reports whether endpoint is a syntactically valid
// https: URL whose host is one of allowedPushHosts or a subdomain of one.
// Port of apps/api/src/routes/push.ts's isAllowedPushEndpoint: the TS
// version anchors its regex on `(^|\.)host$`, i.e. an exact match or a
// match preceded by a dot — never a plain substring check, which is why
// "fcm.googleapis.com.evil.com" (the allowed host as a PREFIX of the real,
// malicious one) does not match here either.
func isAllowedPushEndpoint(endpoint string) bool {
	u, err := url.Parse(endpoint)
	if err != nil || u.Scheme != "https" {
		return false
	}
	host := u.Hostname()
	for _, allowed := range allowedPushHosts {
		if host == allowed || strings.HasSuffix(host, "."+allowed) {
			return true
		}
	}
	return false
}

// GetPushConfig implements GET /api/push/config. REF: "{publicKey}".
func (d Deps) GetPushConfig(_ context.Context, _ gen.GetPushConfigRequestObject) (gen.GetPushConfigResponseObject, error) {
	return gen.GetPushConfig200JSONResponse{PublicKey: d.VAPIDPublicKey}, nil
}

// SubscribePush implements POST /api/push/subscribe. REF: "{endpoint,
// p256dh, auth} → {ok:true}; 400 BAD_ENDPOINT unless https and the host is
// allowlisted. Upserts by endpoint — re-subscribing rebinds it to the
// calling user/family and refreshes its keys" (UpsertPushSubscription's
// ON CONFLICT ("endpoint") DO UPDATE).
func (d Deps) SubscribePush(ctx context.Context, req gen.SubscribePushRequestObject) (gen.SubscribePushResponseObject, error) {
	fam := middleware.FamilyFromContext(ctx)
	if req.Body == nil {
		return nil, errNoRequestBody("SubscribePush")
	}
	body := req.Body

	if !isAllowedPushEndpoint(body.Endpoint) {
		return gen.SubscribePush400JSONResponse{Error: "Unrecognized push service", Code: "BAD_ENDPOINT"}, nil
	}

	if err := d.Q.UpsertPushSubscription(ctx, dbgen.UpsertPushSubscriptionParams{
		FamilyID: fam.FamilyID,
		UserID:   fam.UserID,
		Endpoint: body.Endpoint,
		P256dh:   body.P256dh,
		Auth:     body.Auth,
	}); err != nil {
		return nil, err
	}
	return gen.SubscribePush200JSONResponse{Ok: gen.OkOkTrue}, nil
}

// UnsubscribePush implements POST /api/push/unsubscribe. REF: "{endpoint} →
// {ok:true} (deletes only OWN rows — user-scoped)". Removing an endpoint
// that isn't the caller's, or doesn't exist at all, is still 200: this is a
// set-membership operation ("make sure this endpoint isn't registered to
// me"), not a lookup that can 404.
func (d Deps) UnsubscribePush(ctx context.Context, req gen.UnsubscribePushRequestObject) (gen.UnsubscribePushResponseObject, error) {
	fam := middleware.FamilyFromContext(ctx)
	if req.Body == nil {
		return nil, errNoRequestBody("UnsubscribePush")
	}
	if _, err := d.Q.DeletePushSubscriptionForUser(ctx, dbgen.DeletePushSubscriptionForUserParams{
		Endpoint: req.Body.Endpoint,
		UserID:   fam.UserID,
	}); err != nil {
		return nil, err
	}
	return gen.UnsubscribePush200JSONResponse{Ok: gen.OkOkTrue}, nil
}

// TestPush implements POST /api/push/test. REF: "{sent: n} via
// Deps.Push.ToUser(currentUser)". The payload matches
// apps/api/src/routes/push.ts's byte-for-byte (title, body, and the /home
// deep link the frontend's service worker reads on notification click).
func (d Deps) TestPush(ctx context.Context, _ gen.TestPushRequestObject) (gen.TestPushResponseObject, error) {
	fam := middleware.FamilyFromContext(ctx)
	sent, err := d.Push.ToUser(ctx, fam.UserID, push.PushPayload{
		Title: "Pjokk",
		Body:  "Push works on this device ✅",
		URL:   "/home",
	})
	if err != nil {
		return nil, err
	}
	return gen.TestPush200JSONResponse{Sent: sent}, nil
}
